pandas/tests/indexes/common.py

```diff
@@ -378,6 +378,82 @@ def test_duplicated(self, indices, keep):
         result = idx.duplicated(keep=keep)
         tm.assert_numpy_array_equal(result, expected)
 
+    @pytest.mark.parametrize('keep', ['first', 'last'])
+    def test_duplicated_inverse(self, indices, keep):
+        # check that return_inverse kwarg does not affect outcome
+        if type(indices) is not self._holder:
+            pytest.skip('Can only check if we have the correct type')
+        if not len(indices) or isinstance(indices, MultiIndex):
+            # MultiIndex tested separately in:
+            # tests/indexes/multi/test_unique_and_duplicates
+            pytest.skip('Skip check for empty Index and MultiIndex')
+
+        idx = self._holder(indices)
+        if idx.has_duplicates:
+            # We need to be able to control creation of duplicates here
+            # This is slightly circular, as drop_duplicates depends on
+            # duplicated, but in the end, it all works out because we
+            # cross-check with Series.duplicated
+            idx = idx.drop_duplicates()
+
+        n, k = len(idx), 10
+        duplicated_selection = np.random.choice(n, k * n)
+        idx = self._holder(idx.values[duplicated_selection])
+
+        expected_isdup = idx.duplicated(keep=keep)
+        if keep == 'first':
+            _, tmp_ind, tmp_inv = np.unique(idx, return_index=True,
+                                            return_inverse=True)
+        else:  # 'last'
+            # switch order before calling unique then restore correct ordering
+            # for tmp_ind, tmp_inv
+            _, tmp_ind, tmp_inv = np.unique(idx[::-1], return_index=True,
```

Comment:
Is there no way to simplify this? Pretty difficult to tell what is going on with all of the reverse stepping